### ISSUE TITLE
Engine not work properly when it not set as default.

### ISSUE THREAD

**Issue by AdrianPeniak** [openssl/openssl#4522]
## I did this
I have upgraded from curl version 7.44 to 7.54.1 on my embedded device and my engine does not work properly for https communication anymore, due to segfault in openssl.

## Difference of approach between 7.54.1 and 7.44
```
--- ./src/tool_operate.c
+++ ./src/tool_operate.c
@@ -1296,6 +1296,7 @@
           result = res_setopt_str(curl, CURLOPT_SSLENGINE, config->engine);
           if(result)
             goto show_error;
-       my_setopt(curl, CURLOPT_SSLENGINE_DEFAULT, 1L);
         }
         /* new in curl 7.10.7, extended in 7.19.4. Modified to use
```
As you can see in this version is engine not set as default and this caused segfault. The reason of this change is described in https://github.com/curl/curl/pull/1042 and https://github.com/curl/curl/issues/1973

## curl/libcurl and openssl versions
```
$ curl --version
curl 7.54.1 (arm-poky-linux-gnueabi) libcurl/7.54.1 OpenSSL/1.0.2l zlib/1.2.8
Release-Date: 2017-06-14
Protocols: file ftp ftps http https tftp 
Features: AsynchDNS IPv6 NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy 
```

## debug
```
$ curl -d --sslv3 --cacert /etc/ssl/certs/CA1.pem --cert /var/certs/Cert.pem --key /var/client/key.pem --key-type ENG --engine /usr/lib/libengine.so.0 -H -v https://someHost.com/somethink
Engine Initialization!
*   Trying XXX.XXX.XXX.XXX...
* TCP_NODELAY set
* Connected to someHost.com (XXX.XXX.XXX.XXX) port 443 (#0)
* ALPN, offering http/1.1
engine_load_privkey called
engine_load_privkey: key to load: /var/client/key.pem
engine_load_privkey read_len: 272
private key size is: 264
engine_load_privkey finished
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/CA1.pem
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
Segmentation fault (core dumped)
```

I found that signature proces is different in function RSA_sign (rsa_sign.c). 
For first case (with default engine) is executed fcn _rsa_sign_:
```
 if ((rsa->flags & RSA_FLAG_SIGN_VER) && rsa->meth->rsa_sign) { // rsa->flags(97), RSA_FLAG_SIGN_VER(64), rsa->meth->rsa_sign(0xb6d9dfd0)
        return rsa->meth->rsa_sign(type, m, m_len, sigret, siglen, rsa);
    }
```
For second case (without default engine) is  _rsa_sign_ skipped:
```
if ((rsa->flags & RSA_FLAG_SIGN_VER) && rsa->meth->rsa_sign) { // rsa->flags(38), RSA_FLAG_SIGN_VER(64), rsa->meth->rsa_sign(0)
        return rsa->meth->rsa_sign(type, m, m_len, sigret, siglen, rsa);
    }
```

and process continue with execution _RSA_private_encrypt_ which cause segfault:
```
i = RSA_private_encrypt(i, s, sigret, rsa, RSA_PKCS1_PADDING); // i(51), s(0x3f97f8), sigret(0x3e5cc0), rsa(0x3c0b00), RSA_PKCS1_PADDING(1)
```

Back trace:

```
(gdb) where
#0  0xb6d8db98 in RSA_eay_mod_exp (r0=0x19d3ef4, I=0x19d3ee0, rsa=0x199b440, ctx=0x19d3eb0) at rsa_eay.c:742
#1  0xb6d8ec68 in RSA_eay_private_encrypt (flen=-1227898848, from=<optimized out>, to=0x1 <error: Cannot access memory at address 0x1>, rsa=0x199b440, padding=1) at rsa_eay.c:425
#2  0xb6d8fc30 in RSA_sign (type=<optimized out>, m=m@entry=0xbed181e0 ...., m_len=m_len@entry=64, sigret=sigret@entry=0x19c05e8 ...., siglen=0xbed1818c, siglen@entry=0xbed18184, rsa=rsa@entry=0x199b440) at rsa_sign.c:135
#3  0xb6d9564c in pkey_rsa_sign (ctx=0x19d9db0, sig=0x19c05e8 ...., siglen=0xbed181c4, tbs=0xbed181e0 ...., tbslen=64) at rsa_pmeth.c:271
#4  0xb6dbb148 in EVP_SignFinal (ctx=ctx@entry=0xbed1825c, sigret=sigret@entry=0x19c05e8 ...., siglen=0xbed18250, siglen@entry=0xbed18248, pkey=pkey@entry=0x199ad58) at p_sign.c:104
#5  0xb6e8d860 in ssl3_send_client_verify (s=s@entry=0x19be830) at s3_clnt.c:3298
#6  0xb6e8fe54 in ssl3_connect (s=0x19be830) at s3_clnt.c:465
#7  0xb6e991c0 in ssl23_get_server_hello (s=0x19be830) at s23_clnt.c:802
#8  ssl23_connect (s=0x19be830) at s23_clnt.c:231
#9  0xb6f10368 in ossl_connect_step2 (conn=conn@entry=0x19988c0, sockindex=sockindex@entry=0) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/vtls/openssl.c:2291
#10 0xb6f13ac8 in ossl_connect_common (conn=conn@entry=0x19988c0, sockindex=sockindex@entry=0, nonblocking=nonblocking@entry=true, done=0xb6f13ac8 <ossl_connect_common+164>, done@entry=0xbed1857b) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/vtls/openssl.c:3110
#11 0xb6f14380 in Curl_ossl_connect_nonblocking (conn=conn@entry=0x19988c0, sockindex=sockindex@entry=0, done=done@entry=0xbed1857b) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/vtls/openssl.c:3144
#12 0xb6f14cc4 in Curl_ssl_connect_nonblocking (conn=conn@entry=0x19988c0, sockindex=sockindex@entry=0, done=0xbed1857b) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/vtls/vtls.c:261
#13 0xb6ee2594 in https_connecting (conn=0x19988c0, done=<optimized out>) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/http.c:1408
#14 0xb6f010a4 in multi_runsingle (multi=multi@entry=0x1998460, now=...., data=data@entry=0x1993340) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/multi.c:1610
#15 0xb6f01e1c in curl_multi_perform (multi=multi@entry=0x1998460, running_handles=0xbed18670, running_handles@entry=0xbed18668) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/multi.c:2148
#16 0xb6efa2b4 in easy_transfer (multi=0x1998460) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/easy.c:706
#17 easy_perform (events=false, data=0x1993340) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/easy.c:792
#18 curl_easy_perform (data=0x1993340) at /usr/src/debug/curl/7.54.1-r0/curl-7.54.1/lib/easy.c:811
#19 0x0001ac5c in ?? ()
```
## operating system
embedded linux

**Comment by slandelle** [comment]
Similar issue here.

```
openssl version
OpenSSL 1.0.2n  7 Dec 2017
```

```
openssl s_client -alpn h2 -connect http2.golang.org:443 < /dev/null
CONNECTED(00000003)
140735227965520:error:14077438:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert internal error:s23_clnt.c:802:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 317 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1515690993
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
```

Browsers have no problem connecting to `https://http2.golang.org` over HTTP/2.

**Comment by levitte** [comment]
@slandelle, I have no idea how your case is similar to @AdrianPeniak's, but thanks for bringing this issue back up to our attention.

@AdrianPeniak, your issue actually shows several potential problems, and with regards to keys an engine really has *two* functions:

1.  To be an accelerator.  This is when you would set the engine's RSA implementation as a default, to be used at all times.
2.  To be a key store primarly, and implementing RSA operations *for those keys only*.

In the second case, when a key is loaded via the engine, the engine MUST set the `RSA_FLAG_SIGN_VER` flag on those keys.  Actually, you might as well do that either way.  Otherwise, the libcrypto implementation will be used, and it obviously can't get all the necessary values of the key, and then everything goes belly up.

**Comment by slandelle** [comment]
@levitte I thought so because in both cases error happens at `s23_clnt.c:802`. Sorry if I was mistaken. Should I open a separate issue?

**Comment by levitte** [comment]
@slandelle yes you probably should

**Comment by slandelle** [comment]
@levitte Done as #5059

**Comment by petrovr** [comment]
Now I have time to check initial report with e_nss engine.
First I would like to complete Richard Levitte , about point 2 - engine flag ENGINE_FLAGS_NO_REGISTER_ALL (for details see comment in OpenSSL source code).

Let review paret of command line: .... --key /var/client/key.pem --key-type ENG --engine /usr/lib/libengine.so.0 ...
It seems to me report is not for accelerator engine(1) but for "key store primary" (2) - on line we see arguments --key-type and --key.

a) When I try to reproduce issue I note that curl does not load engine specified by path.
$ curl --version
curl 7.57.0 (x86_64-pc-linux-gnu) libcurl/7.57.0 OpenSSL/1.0.2n zlib/1.2.8 libssh2/1.7.0
Release-Date: 2017-11-29
For --engine argument I have to use engine name. My arguments are : ...
 --key 'SSH RSA test certificate(rsa_sha1) - SSH Test Team cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω'  --key-type ENG  --engine e_nss ...


b) Next I try to initialize engine using OPENSSL_CONF environment variable to specify configuration - no success . This look like defect in "curl" source code!
This is because e_nss engine require configuration parameter that point to "...the directory where the certificate, key, and module databases live..."

c) After modification in e_nss source code - work around for b) I  successfully perform secure connection with following command:
E_NSS_CONFIG_DIR=/work/PKIXSSH/.tests/ssh-nss/test-1.0.2_stable/nssdb \
OPENSSL_ENGINES=/work/PKIXSSH/.tests/ssh-nss/test-1.0.2_stable \
curl -d --sslv3 \
 --cacert /etc/ssl/demoCA/cacert.pem \
 --cert test-1.0.2_stable/keys/rsacrt-client.pem \
 --key 'SSH RSA test certificate(rsa_sha1) - SSH Test Team cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω' \
 --key-type ENG \
 --engine e_nss \
  -H -v \
  https://www.example.net/manual/

Remark e_nss is  build with enabled execution trace log.

Some lines from session:
...
> CONNECT www.example.net:443 HTTP/1.1
...
< HTTP/1.1 200 Connection established
< 
* Proxy replied 200 to CONNECT request
...
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
} [5 bytes data]
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
....
* TLSv1.2 (IN), TLS handshake, Finished (20):
{ [16 bytes data]
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
....
*  SSL certificate verify ok.
....
> POST /manual/ HTTP/1.1
....
</body></html>...TRACE_E_NSS nss_rsa_finish:
....
...TRACE_E_NSS nss_finish() NSS_Shutdown()

....*** Error in `curl': munmap_chunk(): invalid pointer: 0x00007ffe144ecec4 ***
....


To summarize:
A) perhaps libengine.so has to set flag ENGINE_FLAGS_NO_REGISTER_ALL .

If engine methods are not usable as default methods then use of non-engine keys may trigger error (crash) in sign/verify operation.

Is report we see "* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
Segmentation fault (core dumped) "
Perhaps is due to sign operation with non engine key.


B) curl does not initialize/shutdown properly openssl
For sure Curl_ossl_cleanup() is not correct. Initialization Curl_ossl_init() require more detailed review.


In conclusion: 
issue #4522 is is not an openssl defect.

**Comment by jay** [comment]
> B) curl does not initialize/shutdown properly openssl
> For sure Curl_ossl_cleanup() is not correct. Initialization Curl_ossl_init() require more detailed review.
> 
> In conclusion:
> issue #4522 is is not an openssl defect.

If you think you have found a bug in curl please [open an issue](https://github.com/curl/curl/issues/new). I don't understand your analysis but hopefully someone else on the curl team will.
